doc.go

```diff
@@ -0,0 +1,107 @@
+// Generated by running
+//  go generate github.com/gonum/matrix
+// DO NOT EDIT.
+
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package matrix provides common error handling mechanisms for matrix operations
+// in mat64 and cmat128.
+//
+// Overview
+//
+// This section provides a quick overview of the matrix package. The following
+// sections provide more in depth commentary.
+//
+// matrix provides:
+//  - Error type definitions
+//  - Error recovery mechanisms
+//  - Common constants used by mat64 and cmat128
+//
+// Errors
+//
+// The mat64 and cmat128 matrix packages share a common set of errors
+// provided by matrix via the matrix.Error type.
+//
+// Errors are either returned directly or used as the parameter of a panic
+// depending on the class of error encountered. As is usually the case in
+// Go code, returned errors indicate that a call was not able to complete
+// successfully while panics generally indicate a programmer error or an
+// unrecoverable error.
+//
+// Examples of each type are found in the mat64 Solve methods, which find
+// x such that A*x = b.
+//
+// An error value is returned from the function or method when the operation
+// can meaningfully fail. The Solve operation is cannot complete if A is
+// singular. However, determining the singularity of A is most easily
+// discovered during the Solve procedure itself and is a valid result from
```

Comment:
This is where my biology background (and interaction with a theoretical physicist for more than a decade comes in). Here the validity is the system behaviour, not the solution. Note that the text does not say that the solution is valid, but that the result is. This may need clarification, but the result is the ensemble of changes to `m` and the returned `error` - together in the Go/gonum context the result of a call to `m.Solve(a, b)` where a represents a singular matrix is perfectly valid - the result need to be interpreted as "The A matrix is singular and this is how badly, so I cannot solve this for you." This is a valid result.